**Ex 0.1.2: Calculate the value of the following expressions:**
1. The sum of the numbers from 1 to 10
2. The area of a triangle with base length 10 cm and height 5 cm
3. The perimeter of a rectangle of width 20 cm and height 10 cm
4. The amount of interest paid on a €100 balance in one year with an interest rate of 2%
5. The sum of the first ten squares (i.e., $1^2 + 2^2 + \cdots + 10^2$)
6. The amount of interest paid on a €100 balance in 7 years with an interest rate of 2%

In [1]:
# 1. The sum of the numbers from 1 to 10
1+2+3+4+5+6+7+8+9+10     # There's also a compact formula: (10*11)/2

55

In [2]:
# 2. The area of a triangle with base length 10 cm and height 5 cm
(10*5)/2

25

In [3]:
# 3. The perimeter of a rectangle of width 20 cm and height 10 cm
20+10+20+10   # or 2*(20+10) or 2*20 + 2*10

60

In [4]:
# 4. The amount of interest paid on a €100 balance in one year with an interest rate of 2%
100 * 0.02   # Could also do this: 100 * (1.02**1 - 1)

2.0

In [5]:
# 5. The sum of the first ten squares (i.e., $1^2 + 2^2 + \cdots + 10^2$)
1**2 + 2**2 + 3**2 + 4**2 + 5**2 + 6**2 + 7**2 + 8**2 + 9**2 + 10**2
# There's also a compact formula: 10*(10+1)*(2*10+1)/6

385

In [6]:
# 6. The amount of interest paid on a €100 balance in 7 years with an interest rate of 2%
100 * (1.02**7 - 1)

14.868566764928005

**Ex 0.1.3. Calculate the sum of the numbers 1 to 5, showing each intermediate result**
```
>>> 1
1
>>> _+2
3
>>> _+3
6
>>> _+4
10
>>> _+5
15
```

**Ex 0.6.1. How would you test if the value in `a` is odd?**

In [7]:
a = 55
(a % 2) == 1

True

In [8]:
a = 12
(a % 2) == 1

False

In [9]:
a / 2

6

In [10]:
a / 2.0

6.0

In [11]:
(a/2) != (a/2.0)

False

**Ex 0.7.1. Classify the variable `a` as either odd or, if even, a multiple of 4 or not**  
For example, 1 and 3 would be `"odd"`, 4 would be an `"even and a multiple of 4"` and 6 would be `"even but not a multiple of 4"`

In [12]:
a = 6
if (a % 2) == 1:
    description = "odd"
else:
    if (a % 4) == 0:
        description = "even and a multiple of 4"
    else:
        description = "even but not a multiple of 4"

"a is " + description

'a is even but not a multiple of 4'

**Ex 0.7.2. Rewrite your solution to Ex 0.7.1 more concisely using `elif`**

In [13]:
a = 6
if (a % 2) == 1:
    description = "odd"
elif (a % 4) == 0:
    description = "even and a multiple of 4"
else:
    description = "even but not a multiple of 4"

"a is " + description

'a is even but not a multiple of 4'

**Ex 0.7.3. Write a loop to calculate `10! = 10 * 9 * 8 * ... * 1`**

In [14]:
# Loop from 1 to 10
i = 1
result = 1
while i <= 10:
    result = result * i
    i = i + 1

"10! = " + str(result)

'10! = 3628800'

In [15]:
# Loop from 10 to 1
i = 10
result = 1
while i >= 1:
    result = result * i
    i = i - 1

"10! = " + str(result)

'10! = 3628800'

**Ex 0.7.4. Implement Newton's algorithm to find the square root of 2 to 5 decimal places**

In [16]:
n = 2.0
x = 1.0
eps = 1e-5
while not (n - eps) < (x*x) < (n + eps):
    x = (x + n / x) / 2

"The square root of " + str(n) + " is " + str(x) + " +/- " + str(eps)

'The square root of 2.0 is 1.41421568627 +/- 1e-05'

**(!) Ex 0.7.5. Modify your code for Newton's method to show the value of the guess after each iteration, as well as the total number of iterations**

In [17]:
n = 2.0
x = 1.0
eps = 1e-10
i = 0
print "Starting with x = " + str(x)
while not (n - eps) < (x*x) < (n + eps):
    x = (x + n / x) / 2
    i = i + 1
    print("After " + str(i) + " iteration(s), x = " + str(x))
    
"The square root of " + str(n) + " is " + str(x) + " +/- " + str(eps)

Starting with x = 1.0
After 1 iteration(s), x = 1.5
After 2 iteration(s), x = 1.41666666667
After 3 iteration(s), x = 1.41421568627
After 4 iteration(s), x = 1.41421356237


'The square root of 2.0 is 1.41421356237 +/- 1e-10'

**(!) Ex 0.7.6. Write a loop to calculate the Collatz sequence for any number `n`.  Try it for `n=13` and `n=25`.**

In [18]:
n = 25
print "Starting with: " + str(n)
while n != 1:
    if (n % 2) == 0:  # even
        n = n / 2
    else:
        n = 3*n + 1
    print "Now at " + str(n)


Starting with: 25
Now at 76
Now at 38
Now at 19
Now at 58
Now at 29
Now at 88
Now at 44
Now at 22
Now at 11
Now at 34
Now at 17
Now at 52
Now at 26
Now at 13
Now at 40
Now at 20
Now at 10
Now at 5
Now at 16
Now at 8
Now at 4
Now at 2
Now at 1


**Ex 0.8.1.  Rewrite your code for Collatz sequences to produce a list instead of printing the items in the sequence**

In [19]:
n = 25
result = [n]
while n != 1:
    if (n % 2) == 0:  # even
        n = n / 2
    else:
        n = 3*n + 1
    result.append(n)

print str(result)

[25, 76, 38, 19, 58, 29, 88, 44, 22, 11, 34, 17, 52, 26, 13, 40, 20, 10, 5, 16, 8, 4, 2, 1]


**Ex 0.8.2. Calculate the sum of the numbers from 1 to 10 using a `for` loop with `range`**

In [20]:
total = 0
for i in range(1, 10+1):
    total = total + i
total

55

**Ex 0.8.3. Calculate the sum of the numbers from 1 to 10 using `sum` and `range`**

In [21]:
sum(range(1,10+1))

55

**Ex 0.8.4. How would you calculate the average of the value in the list?**

In [22]:
l = [3, 5, 72]
sum(l) / float(len(l))  # Don't forget float(...)!
                        # Otherwise you'll get truncating division
                        # for integers

26.666666666666668

**Ex 0.8.5. First, without list comprehensions, build a list of squares from $1^2$ to $10^2$ using a `for` loop and `range`**

In [23]:
squares = []
for i in range(1,10+1):
    squares.append(i*i)
print(str(squares))

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


**Ex 0.8.6. Using list comprehension, make a list of all divisors of a number `n`, e.g. for n = 6, the list would be [1,2,3,6]**

In [24]:
n = 24
[d for d in range(1,n+1) if (n % d) == 0]

[1, 2, 3, 4, 6, 8, 12, 24]

**Ex 0.9.1.  How would you classify a filename based on its extension as either a JPEG, a text file, a CSV, or something else?**

In [25]:
filename = "myfile.csv"
if filename[-4:] == ".jpg":
    filetype = "JPEG"
elif filename[-4:] == ".txt":
    filetype = "text file"
elif filename[-4:] == ".csv":
    filetype = "CSV file"
else:
    filetype = "something else"
    
print(filename + ": " + filetype)

myfile.csv: CSV file


**Ex 0.9.2. Given a string for a phone number, e.g. "+32-123-456789", count the number of hyphens in it**

In [26]:
phone = "+32-123-456789"
count = 0
for c in phone:
    if c == '-':
        count = count + 1
        
print(phone + " has " + str(count) + " hyphens")

+32-123-456789 has 2 hyphens


**Ex 0.9.3. Repeat exercise 0.9.2 using list comprehensions**

In [27]:
phone = "+32-123-456789"
count = sum([1 for c in phone if c == '-'])
print(phone + " has " + str(count) + " hyphens")

+32-123-456789 has 2 hyphens


**Ex 0.9.4. Using `join` and a list comprehension, remove anything except the digits in the phone number "+32-123-456789"**

In [28]:
phone = "+32-123-456789"
"".join([c for c in phone if c.isdigit()])

'32123456789'

**Ex 0.9.5. Using `split`, `strip` and a list comprehension, take a line from a CSV and make a list of the fields in there, with leading and trailing spaces removed**

In [29]:
csvline=" one ,  two, three"
[f.strip() for f in csvline.split(",")]

['one', 'two', 'three']

*Optional:* Given `csvline="  1, 2.4,5, 1000"`, could you turn this into a list of `float`s?

In [30]:
csvline="  1, 2.4,5, 1000"
[float(field.strip()) for field in csvline.split(",")]

[1.0, 2.4, 5.0, 1000.0]

**Ex 0.9.6. Look up the help string for `split`.  Now split the string "one,two,three" into two parts, one before the first command and then everything after that**

In [31]:
s = "one,two,three"
s.split(",", 1)

['one', 'two,three']

**Ex 0.9.7. Look up the help string for `replace`.  Now replace every hyphen in the phone number "+32-123-456789" with a dot (`'.'`)**

In [32]:
phone = "+32-123-456789"
phone.replace("-", ".")

'+32.123.456789'

**Ex 0.10.1. Using list comprehension, take a number `n` and make a list of all the ways of expressing it as a product of two positive integers.**  
For example, given 8, your list should read [(1,8), (2,4), (4,2), (8,1)]

In [33]:
# This is the straightforward solution; you can optimize this a lto
n = 24
[(a, b) for a in range(1,n+1) for b in range(1,n+1) if a*b == n]

[(1, 24), (2, 12), (3, 8), (4, 6), (6, 4), (8, 3), (12, 2), (24, 1)]

In [34]:
n = 24
l = []
for a in range(1,n+1):
    for b in range(1,n+1):
        if a*b == n:
            l.append((a, b))
print l

[(1, 24), (2, 12), (3, 8), (4, 6), (6, 4), (8, 3), (12, 2), (24, 1)]


**(!) Ex 0.10.2. Given a list `l`, produce a list of tuples for the consecutive items in `l`.**  
For example, given `l = ['a', 'g', 'x']`, your code would produce the list `[('a', 'g'), ('g', 'x')]`  
_Hint_: You'll need to be creative with `zip` and slicing

In [35]:
l = ['a', 'g', 'x']
zip(l[:-1], l[1:])

[('a', 'g'), ('g', 'x')]

**(!) Ex 0.11.1  Write some code to create a dictionary that maps each word in a sentence to the number of times it appears in it.  For simplicity, assume words are only separated by spaces.**  
For example, given the sentence "I love love love New York but I don't love York", your code would output the dictionary
```
{'I': 2, 'love': 4, 'York': 2, 'but': 1, "don't": 1, "New": 1}
```

In [36]:
sentence = "I love love love New York but I don't love York"
result = {}
for word in sentence.split(" "):
    if word in result:
        result[word] = result[word] + 1
    else:
        result[word] = 1
        
result

{'I': 2, 'New': 1, 'York': 2, 'but': 1, "don't": 1, 'love': 4}

An alternate solution that works for small inputs but will struggle with larger inputs is given below:

In [37]:
sentence = "I love love love New York but I don't love York"
result = {}
words = sentence.split(" ")
for word in words:
    if word not in result:
        result[word] = words.count(word)
print result

{'love': 4, "don't": 1, 'I': 2, 'but': 1, 'York': 2, 'New': 1}


**Ex 0.12.1  Using functions in `math`, calculate $\sin(30^\circ)$, $\sqrt(5)$ and $e^{2.5}$**

In [38]:
import math
print("sin(30 degrees) = {0}".format(math.sin(math.radians(30))))
print("sqrt(5) = {0}".format(math.sqrt(5)))
print("e^2.5 = {0}".format(math.e ** 2.5))
print("e^2.5 = {0}".format(math.exp(2.5)))  # Same as above

sin(30 degrees) = 0.5
sqrt(5) = 2.2360679775
e^2.5 = 12.1824939607
e^2.5 = 12.1824939607


**Ex 0.12.2 Create a script `squaresum.py` that takes a number `n` as parameter and prints out the sum of the first `n` squares.**  
For example, running `python squaresum.py 10` prints out `385`

```
import sys
if len(sys.argv) != 2:
    print("Usage: python squaresum.py n")
else:
    _, nstr = sys.argv
    n = int(nstr)
    print(sum([i*i for i in range(1, n+1)]))
```

**Ex 0.13.1.  Previously, we wrote some code to count the frequencies of words in a string.  Turn it into a function that you can invoke like this: freqs_of_words("I love love love New York")**

In [39]:
def freqs_of_words(sentence):
    result = {}
    for word in sentence.split(" "):
        if word in result:
            result[word] = result[word] + 1
        else:
            result[word] = 1
    return result

In [40]:
freqs_of_words("I love love love New York")

{'I': 1, 'New': 1, 'York': 1, 'love': 3}

**(!) Ex 0.13.2. Write a function to calculate the $n$th Fibonacci number, defined recursively as follows:**

$$
F(n) = F(n-1) + F(n-2).
$$

**It's customary to specify $F(0)$ and $F(1)$ explicitly.  Accept keyword arguments for these edge cases, with a default value of $1$ for both.**

**What's the value of $F(8)$ when $F(0) = F(1) = 1$?  What about when $F(0) = 10$ and $F(1) = 15$?**

In [41]:
def fib(n, fib_zero=1, fib_one=1):
    if n == 0:
        return fib_zero
    elif n == 1:
        return fib_one
    else:
        return fib(n-1, fib_zero, fib_one) + fib(n-2, fib_zero, fib_one)

In [42]:
fib(8)

34

In [43]:
fib(8, fib_zero=10, fib_one=15)

445